sdk/src/main/java/com/qonversion/android/sdk/QonversionRepository.kt

```diff
@@ -122,127 +128,216 @@ internal class QonversionRepository private constructor(
 
         api.purchase(purchaseRequest).enqueue {
             onResponse = {
                 logger.log("purchaseRequest - success - $it")
-                callback?.onSuccess(storage.load())
-                kickRequestQueue()
+                handlePermissionsResponse(it, callback)
             }
             onFailure = {
-                logger.log("purchaseRequest - failure - $it")
+                logger.log("purchaseRequest - failure - ${it?.toQonversionError()}")
                 if (it != null) {
-                    callback?.onError(it)
+                    callback.onError(it.toQonversionError())
                 }
             }
         }
     }
 
+    private fun convertPurchases(purcahses: List<Purchase>?): List<Inapp> {
+        val inapps: MutableList<Inapp> = mutableListOf()
+
+        purcahses?.forEach {
+            val inapp = convertPurchase(it)
+            inapps.add(inapp)
+        }
+
+        return inapps.toList()
+    }
+
+    private fun convertPurchase(purchase: Purchase): Inapp {
+        val purchaseDetails = convertPurchaseDetails(purchase)
+        val introductoryOfferDetails = convertIntroductoryPurchaseDetail(purchase)
+
+        return Inapp(purchaseDetails, introductoryOfferDetails)
+    }
+
+    private fun convertIntroductoryPurchaseDetail(purchase: Purchase): IntroductoryOfferDetails? {
+        var introductoryOfferDetails: IntroductoryOfferDetails? = null
+
+        if ((purchase.freeTrialPeriod.isNotEmpty() || purchase.introductoryAvailable)
+            && purchase.introductoryPeriodUnit != null
+            && purchase.introductoryPeriodUnitsCount != null) {
+            introductoryOfferDetails = IntroductoryOfferDetails(
+                purchase.introductoryPrice,
+                purchase.introductoryPeriodUnit,
+                purchase.introductoryPeriodUnitsCount,
+                purchase.introductoryPriceCycles,
+                purchase.paymentMode
+            )
+        }
+
+        return introductoryOfferDetails
+    }
+
+    private fun convertPurchaseDetails(purchase: Purchase): PurchaseDetails {
+        val purchaseDetail = PurchaseDetails(
+            purchase.productId,
+            purchase.purchaseToken,
+            purchase.purchaseTime,
+            purchase.priceCurrencyCode,
+            purchase.price,
+            purchase.orderId,
+            purchase.originalOrderId,
+            purchase.periodUnit,
+            purchase.periodUnitsCount,
+            null
+        )
+
+        return purchaseDetail
+    }
+
+    private fun convertHistory(historyRecords: List<PurchaseHistoryRecord>): List<History> {
+        val histories: MutableList<History> = mutableListOf()
+
+        historyRecords.forEach {
+            val history = History(
+                it.sku,
+                it.purchaseToken,
+                it.purchaseTime.milliSecondsToSeconds()
+            )
+            histories.add(history)
+        }
+
+        return histories.toList()
+    }
```

Comment:
It would be the same, but shorter:
val histories=historyRecords.map { History(
                it.sku,
                it.purchaseToken,
                it.purchaseTime.milliSecondsToSeconds()
            ) }